### PR TITLE
feat: Add generic OAUTH/OIDC support WPB-2053

### PR DESCRIPTION
**Pull request by KaterinaWire**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-2053" title="WPB-2053" target="_blank">StoryWPB-2053</a>  [iOS] Add generic OAUTH/OIDC support in client
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

This PR introduces a new OAuth Use Case that provides client authentication to obtain an idToken from an identity provider. This use case is also injected to the certificate enrollment flow.
One of the requirements was to use the AppAuth SDK.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+                return continuation.resume(returning: request)
+            }
+        }
+
+        return try await execute(authorizationRequest: request)
+    }
+
+    private func execute(authorizationRequest: OIDAuthorizationRequest) async throws -> IdToken {
+        guard let userAgen = OIDExternalUserAgentIOS(presenting: rootViewController) else {
```
Comment (typfel) [inline wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift]:
```suggestion
        guard let userAgent = OIDExternalUserAgentIOS(presenting: rootViewController) else {
```

File: wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+            throw OAuthError.failedToRetrieveConfiguration
+        }
+
+        let request: OIDAuthorizationRequest = try await withCheckedThrowingContinuation { continuation in
+            OIDAuthorizationService.discoverConfiguration(forIssuer: identityProvider) { configuration, error in
+                guard
+                    let config = configuration,
+                    error == nil
+                else {
+                    return continuation.resume(throwing: OAuthError.failedToRetrieveConfiguration)
```
Comment (typfel) [inline wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift]: Let's include the underlaying error here?

File: wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                return continuation.resume(returning: request)
+            }
+        }
+
+        return try await execute(authorizationRequest: request)
+    }
+
+    private func execute(authorizationRequest: OIDAuthorizationRequest) async throws -> IdToken {
+        guard let userAgen = OIDExternalUserAgentIOS(presenting: rootViewController) else {
+            throw OAuthError.missingOIDExternalUserAgen
```
Comment (typfel) [inline wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift]:
```suggestion
            throw OAuthError.missingOIDExternalUserAgent
```

File: wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        }
+
+        return try await withCheckedThrowingContinuation { [weak self] continuation in
+            self?.currentAuthorizationFlow = OIDAuthState.authState(byPresenting: authorizationRequest,
```
Comment (typfel) [inline wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift:83]: Why do we need store this? We don't seem to access it

File: wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                                                                   externalUserAgent: userAgen,
+                                                                   callback: { authState, error in
+                guard error == nil else {
+                    return continuation.resume(throwing: OAuthError.failedToSendAuthorizationRequest)
```
Comment (typfel) [inline wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift]: Include the underlaying error

File: wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                }
+
+                authState?.performAction { (_, idToken, error) in
+                    guard error == nil else {
+                        return continuation.resume(throwing: OAuthError.failedToSendAuthorizationRequest)
```
Comment (typfel) [inline wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift]: Include the underlaying error

File: wire-ios-request-strategy/Sources/E2EIdentity/EnrollE2eICertificateUseCase.swift [hunk 7]
```diff
@@ -33,37 +39,40 @@ public final class EnrollE2eICertificateUseCase: EnrollE2eICertificateUseCaseInt
...
-            let e2eiClientId = E2eIClientID(user: selfUser)
-        else {
-            throw EnrollE2EICertificateUseCaseFailure.failedToSetupEnrollment
-        }
-
-        let enrollment = try await e2eiRepository.createEnrollment(e2eiClientId: e2eiClientId, userName: userName, handle: handle)
+    public func invoke(e2eiClientId: E2eIClientID,
+                       userName: String,
+                       userHandle: String,
+                       invokeAuthentication: OAuthBlock) async throws -> String {
```
Comment (typfel) [inline wire-ios-request-strategy/Sources/E2EIdentity/EnrollE2eICertificateUseCase.swift]: why not use a nice verb instead

```suggestion
                       authenticate: OAuthBlock) async throws -> String {
```

**Review by typfel — CHANGES_REQUESTED** [review]

**Review by caldrian — APPROVED** [review]

File: wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift (hunk 4 above)
Comment (KaterinaWire) [reply wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift:83]: I need to have a property in order to continue the authorization flow from the redirect:
https://openid.github.io/AppAuth-iOS/

File: wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift (hunk 5 above)
Comment (KaterinaWire) [reply wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift]: done

File: wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift (hunk 2 above)
Comment (KaterinaWire) [reply wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift]: done

File: wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift (hunk 6 above)
Comment (KaterinaWire) [reply wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift]: done

**Review by typfel — APPROVED** [review]

**Pull request merged**